clients/include/host_batch_vector.hpp

```diff
@@ -0,0 +1,191 @@
+//
+// Copyright 2018-2019 Advanced Micro Devices, Inc.
+//
+#pragma once
+
+#include <string.h>
+
+//!
+//! @brief Implementation of the batch vector on host.
+//!
+template <typename T>
+class host_batch_vector
+{
+public:
+    //!
+    //! @brief Delete copy constructor.
+    //!
+    host_batch_vector(const host_batch_vector<T>& that) = delete;
+
+    //!
+    //! @brief Delete copy assignement.
+    //!
+    host_batch_vector& operator=(const host_batch_vector<T>& that) = delete;
+
+    //!
+    //! @brief Constructor.
+    //! @param size_vector The size of one of the vectors.
+    //! @param batch_count The number of vectors.
+    //!
+    explicit host_batch_vector(size_t size_vector, rocblas_int batch_count) noexcept
+        : m_batch_count(batch_count)
+        , m_size(size_vector)
+    {
+        if(false == this->try_initialize_memory())
+        {
+            this->free_memory();
+        }
+    };
+
+    //!
+    //! @brief Destructor.
+    //!
+    ~host_batch_vector() noexcept
+    {
+        this->free_memory();
+    };
+
+    //!
+    //! @brief Returns the size of the vector.
+    //!
+    inline size_t size() const noexcept
+    {
+        return this->m_size;
+    };
+
+    //!
+    //! @brief Returns the batch count.
+    //!
+    inline rocblas_int batch_count() const noexcept
+    {
+        return this->m_batch_count;
+    };
+
+    //!
+    //! @brief Random access to the vectors.
+    //! @param batch_index the batch index.
+    //! @return The mutable pointer.
+    //!
+    inline T* operator[](rocblas_int batch_index) noexcept
+    {
+        assert(0 <= batch_index && this->m_batch_count > batch_index);
+        return this->m_data[batch_index];
+    };
+
+    //!
+    //! @brief Constant random access to the vectors.
+    //! @param batch_index the batch index.
+    //! @return The non-mutable pointer.
+    //!
+    inline const T* operator[](rocblas_int batch_index) const noexcept
+    {
+        assert(0 <= batch_index && this->m_batch_count > batch_index);
+        return this->m_data[batch_index];
+    };
+
+    //!
+    //! @brief Cast to a double pointer.
+    //!
+    inline operator T**() noexcept
+    {
+        return this->m_data;
+    }
+
+    //!
+    //! @brief Constant cast to a double pointer.
+    //!
+    inline operator const T* const*() noexcept
+    {
+        return this->m_data;
+    }
+
+    //!
+    //! @brief Copy from a host batched vector.
+    //! @param that the vector the data is copied from.
+    //! @return true if the copy is done successfully, false otherwise.
+    //!
+    bool copy_from(const host_batch_vector<T>& that) noexcept
```

Comment:
Wouldn't operator= be acceptable here?

Reply:
Originally. assignment operators have been  specified as 'deleted'.
Moreover, it's been written in this way to be able to catch that something wrong is happening during the copying operation without throwing an exception.

Reply:
[Rule of 3/5/0](https://en.cppreference.com/w/cpp/language/rule_of_three)

Assignment operator was intentionally deleted in the original implementation. Better to just get a compilation error than to deal with runtime complications. 

We could define it like a "copy" function if we wanted to, so long as it was well-defined and handled different sizes, etc.

Reply:
For now, the behavior is performing the copy if all the parameters are the same.
We can start with that and extend if needed.

copy_from can return a flag to indicate if the copy has been performed successfully or not.
An assignment operator cannot.